src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -289,15 +296,32 @@ def backfill_artifact_index_updates():
                 redis_client.srem(deletion_key, *bundles_to_remove)
 
     # Then, we process any pending removals marked in redis.
-    deletion_keys = redis_client.scan_iter(
-        match=DELETION_KEY_PREFIX + ":*",
-    )
-    for deletion_key in deletion_keys:
-        _prefix, idx_id = deletion_key.split(":")
-
+    # NOTE on the usage of redis sets:
+    # We could use a redis `SCAN` to find all the sets of scheduled deletions,
+    # however that scales with the *total* number of keys in the redis server.
+    # Therefore, we rather use a second redis set for the indexes that have deletions scheduled.
+    # Races when adding to these sets should not cause a problem, as the *writer*
+    # side first adds to the per-index set, and then adds the index to the set of indexes
+    # needing deletions.
+    # However, there is indeed a slight consistency problem here in case we hit an
+    # error after removing the index from the set.
+    # In that case we would indeed forget that this index has pending removals,
+    # until another deletion request re-adds it to the set of indexes.
+    # If this is indeed a problem, we can try to indeed use `scan_iter` to find all
+    # the indexes that were missed, like so:
+    # deletion_keys = redis_client.scan_iter(
+    #     match=DELETION_KEY_PREFIX + ":*",
+    # )
+    # for deletion_key in deletion_keys:
+    #     _prefix, idx_id = deletion_key.split(":")
```

Comment:
This is a bit unfortunate. We could as well create two separate tasks, one for updates that runs frequently and is expected to be quick, and one for deletes that runs less frequently and is allowed to be a lot slower and does a thorough scan. Maybe that is the best compromise here 🤷🏻‍♂️